### ISSUE TITLE
Exception when clicking on various hyperlinks (this time users)

### ISSUE THREAD

**Issue by rmr** [FusionAuth/fusionauth-issues#877]
Exception when clicking on various hyperlinks (this time users)

### Description 
This exception occured when clicking on the users hyperlink. However, I have seen when using chrome, and trying to login, I get the same error. When I restart chrome, and login again, I no longer see the exception, until a later time.

I have tried SQL delete the refresh token table, but it didnt help

### Affects versions
1.19.6

### Screenshots
2020-09-18 11:29:46.632 AM ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown    
 java.lang.NullPointerException: null                                                                                                     
     at io.fusionauth.api.service.jwt.DefaultRefreshTokenService.refreshSSOSessionToken(DefaultRefreshTokenService.java:186)              
     at io.fusionauth.app.service.security.DefaultSSOService.login(DefaultSSOService.java:107)                                            
     at io.fusionauth.app.action.oauth2.BaseOAuthAction.handleInteractiveLoginResponse(BaseOAuthAction.java:414)                          
     at io.fusionauth.app.action.oauth2.BaseOAuthAction.authenticateAuthorizationGrant(BaseOAuthAction.java:190)                          
     at io.fusionauth.app.action.oauth2.AuthorizeAction.post(AuthorizeAction.java:146)                                                    
     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)                                                    
     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)                                  
     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)                          
     at java.base/java.lang.reflect.Method.invoke(Method.java:564)                                                                        
     at org.primeframework.mvc.util.ReflectionUtils.invoke(ReflectionUtils.java:414)                                                      
     at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.execute(DefaultActionInvocationWorkflow.java:79)                    
     at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.perform(DefaultActionInvocationWorkflow.java:62)                    
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.validation.DefaultValidationWorkflow.perform(DefaultValidationWorkflow.java:47)                            
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.security.DefaultSecurityWorkflow.perform(DefaultSecurityWorkflow.java:60)                                  
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.parameter.DefaultPostParameterWorkflow.perform(DefaultPostParameterWorkflow.java:50)                       
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.content.DefaultContentWorkflow.perform(DefaultContentWorkflow.java:52)                                     
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.parameter.DefaultParameterWorkflow.perform(DefaultParameterWorkflow.java:57)                               
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.parameter.DefaultURIParameterWorkflow.perform(DefaultURIParameterWorkflow.java:102)                        
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.scope.DefaultScopeRetrievalWorkflow.perform(DefaultScopeRetrievalWorkflow.java:58)                         
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.message.DefaultMessageWorkflow.perform(DefaultMessageWorkflow.java:44)                                     
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at io.fusionauth.app.primeframework.FrontEndTenantWorkflow.perform(FrontEndTenantWorkflow.java:78)                                   
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.action.DefaultActionMappingWorkflow.perform(DefaultActionMappingWorkflow.java:126)                         
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.workflow.StaticResourceWorkflow.perform(StaticResourceWorkflow.java:97)                                    
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.parameter.RequestBodyWorkflow.perform(RequestBodyWorkflow.java:89)                                         
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at org.primeframework.mvc.security.DefaultSavedRequestWorkflow.perform(DefaultSavedRequestWorkflow.java:64)                          
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at io.fusionauth.app.primeframework.CORSFilter.handleNonCORS(CORSFilter.java:291)                                                    
     at io.fusionauth.app.primeframework.CORSFilter.doFilter(CORSFilter.java:223)                                                         
     at io.fusionauth.app.primeframework.CORSRequestWorkflow.perform(CORSRequestWorkflow.java:48)                                         
     at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)                                       
     at io.fusionauth.app.primeframework.FusionAuthMVCWorkflow.perform(FusionAuthMVCWorkflow.java:88)                                     
     at org.primeframework.mvc.workflow.DefaultWorkflowChain.continueWorkflow(DefaultWorkflowChain.java:44)                               
     at org.primeframework.mvc.servlet.FilterWorkflowChain.continueWorkflow(FilterWorkflowChain.java:50)                                  
     at org.primeframework.mvc.servlet.PrimeFilter.doFilter(PrimeFilter.java:78)                                                          
     at com.inversoft.maintenance.servlet.MaintenanceModePrimeFilter.doFilter(MaintenanceModePrimeFilter.java:59)                         
     at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)                                 
     at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)                                         
     at com.inversoft.servlet.UTF8Filter.doFilter(UTF8Filter.java:27)                                                                     
     at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)                                 
     at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)                                         
     at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199)                                               
     at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)                                                
     at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:543)                                            
     at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)                                                     
     at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)                                                      
     at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)                                                  
     at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)                                                       
     at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:615)                                                             
     at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)                                                  
     at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:818)                                           
     at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1626)                                               
     at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)                                                   
     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)                                         
     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)                                         
     at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)                                                
     at java.base/java.lang.Thread.run(Thread.java:832)                                                                                   
 stream closed                                         
### Platform
Kubernetes/Docker image

**Comment by robotdan** [comment]
Thanks for reporting @rmr !

**Comment by robotdan** [comment]
I think I can see how this could happen in theory, but haven't recreated yet. 

Can you clarify your recreate steps? Something like:

1. Log in ok
2. Wait for SSO to timeout
3. With a browser already open? click a link

**Comment by Kjir** [comment]
I encounter the same error while doing a load test. I can provide the locust script that I wrote, but here is the gist of it:

```python
    @task(100)
    def login(self):
        if self.tenant is None or len(self.registered_users) < 1:
            return
        random_user_index = floor(random() * len(self.registered_users))
        user = self.registered_users[random_user_index]["user"]
        found_users = self.client.search_users_by_query(
            {"search": {"queryString": user["email"]}}
        ).success_response
        if found_users["total"] < 1:
            return
        client_id = found_users["users"][0]["registrations"][0]["applicationId"]
        response = self.httpclient.get(
            "/oauth2/authorize",
            allow_redirects=False,
            params={
                "client_id": client_id,
                "redirect_uri": "https://example.com/auth/login",
                "response_type": "code",
                "scope": "offline_access",
                "username": user["email"],
            },
        )
        form_response = self.httpclient.post(
            "/oauth2/authorize",
            allow_redirects=False,
            data={
                "loginId": user["email"],
                "password": "unsafeunsafe",
                "redirect_uri": "https://example.com/auth/login",
                "client_id": client_id,
                "response_type": "code",
            },
        )
        qs = urlparse(form_response.headers["Location"]).query
        self.auth_code = parse_qs(qs)["code"]
```

The error happens on the `POST /oauth2/authorize` call with a load of around 5 requests per second. The 95% percentile response time is still below 200ms, so I find a timeout situation quite unlikely.
![image](https://user-images.githubusercontent.com/64094/93992020-55356480-fd8d-11ea-9777-b6c790ecf94b.png)

**Comment by soullivaneuh** [comment]
Customer are reporting logout crashes after letting being afk for a while.

Here is the error I get, similar to the first report: 

```
ERROR io.fusionauth.app.primeframework.error.ExceptionExceptionHandler - An unhandled exception was thrown
java.lang.NullPointerException: null
	at io.fusionauth.api.service.jwt.DefaultRefreshTokenService.refreshSSOSessionToken(DefaultRefreshTokenService.java:186)
	at io.fusionauth.app.service.security.DefaultSSOService.login(DefaultSSOService.java:107)
	at io.fusionauth.app.action.oauth2.AuthorizeAction.get(AuthorizeAction.java:70)
	at jdk.internal.reflect.GeneratedMethodAccessor85.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.primeframework.mvc.util.ReflectionUtils.invoke(ReflectionUtils.java:414)
	at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.execute(DefaultActionInvocationWorkflow.java:79)
	at org.primeframework.mvc.action.DefaultActionInvocationWorkflow.perform(DefaultActionInvocationWorkflow.java:62)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.validation.DefaultValidationWorkflow.perform(DefaultValidationWorkflow.java:47)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.security.DefaultSecurityWorkflow.perform(DefaultSecurityWorkflow.java:60)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.parameter.DefaultPostParameterWorkflow.perform(DefaultPostParameterWorkflow.java:50)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.content.DefaultContentWorkflow.perform(DefaultContentWorkflow.java:52)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.parameter.DefaultParameterWorkflow.perform(DefaultParameterWorkflow.java:57)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.parameter.DefaultURIParameterWorkflow.perform(DefaultURIParameterWorkflow.java:102)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.scope.DefaultScopeRetrievalWorkflow.perform(DefaultScopeRetrievalWorkflow.java:58)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.message.DefaultMessageWorkflow.perform(DefaultMessageWorkflow.java:44)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at io.fusionauth.app.primeframework.FrontEndTenantWorkflow.perform(FrontEndTenantWorkflow.java:165)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.action.DefaultActionMappingWorkflow.perform(DefaultActionMappingWorkflow.java:126)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.workflow.StaticResourceWorkflow.perform(StaticResourceWorkflow.java:97)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.parameter.RequestBodyWorkflow.perform(RequestBodyWorkflow.java:89)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at org.primeframework.mvc.security.DefaultSavedRequestWorkflow.perform(DefaultSavedRequestWorkflow.java:64)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at io.fusionauth.app.primeframework.CORSFilter.handleNonCORS(CORSFilter.java:291)
	at io.fusionauth.app.primeframework.CORSFilter.doFilter(CORSFilter.java:252)
	at io.fusionauth.app.primeframework.CORSRequestWorkflow.perform(CORSRequestWorkflow.java:48)
	at org.primeframework.mvc.workflow.SubWorkflowChain.continueWorkflow(SubWorkflowChain.java:51)
	at io.fusionauth.app.primeframework.FusionAuthMVCWorkflow.perform(FusionAuthMVCWorkflow.java:88)
	at org.primeframework.mvc.workflow.DefaultWorkflowChain.continueWorkflow(DefaultWorkflowChain.java:44)
	at org.primeframework.mvc.servlet.FilterWorkflowChain.continueWorkflow(FilterWorkflowChain.java:50)
	at org.primeframework.mvc.servlet.PrimeFilter.doFilter(PrimeFilter.java:78)
	at com.inversoft.maintenance.servlet.MaintenanceModePrimeFilter.doFilter(MaintenanceModePrimeFilter.java:59)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at com.inversoft.servlet.UTF8Filter.doFilter(UTF8Filter.java:27)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:543)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:615)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:818)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1626)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

**Comment by robotdan** [comment]
Thanks for the additional details @soullivaneuh and @rmr. I did make one change in 1.19.7 that in theory will fix this - but I have yet to recreate so I don't know for sure. 

If you do end up upgrading to 1.19.7, please report if you still see it. In the mean time we'll attempt to recreate.

**Comment by soullivaneuh** [comment]
@robotdan Thanks, I applied the update. I'll inform you if the error happens again.

**Comment by soullivaneuh** [comment]
@robotdan This specific error does not seems to occurs anymore since `1.19.7`, it looks like you fixed the problem. :+1: 

I'm curious: Why are you not sure about the fix you did?

**Comment by rmr** [comment]
This fix seems to work for me too! I have been swamped here, so i apologize for the slow and non-existent replies.

**Comment by robotdan** [comment]
@soullivaneuh I think I understood why the NPE was occurring, but I have not been able to identify a sequence of steps that would cause it. This means I haven't been able to write a test to prove it could occur, and thus prove my fix worked. 

Unless I can write a test to prove it, I'm always unsure. :-)

**Comment by robotdan** [comment]
I think I'll close this one out as fixed under 1.19.7 since we have two confirmations that it no longer occurs. Thanks for reporting and providing us valuable feedback!

**Comment by robotdan** [comment]
@Kjir feel free to report on this thread as well if the issue has been resolved for you as well.
